File: Cisco/11TopKFrequentElements.py
from collections import Counter


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        counts = Counter(nums)
        max_freq = max(counts.values())
        reversed_counts = [[] for _ in range(max_freq + 1)]

        for num, count in counts.items():
            reversed_counts[count].append(num)

        res = []
        for i in range(max_freq, 0, -1):
            res.extend(reversed_counts[i])
            if len(res) >= k:
                return res[:k]

File: Cisco/GitHubList/C++/11TopKFrequentElements.cpp
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        unordered_map<int, int> counter;
        for (const auto &num : nums)
            counter[num]++;

        priority_queue<pair<int, int>, vector<pair<int, int>>, less<>> maxHeap;

        for (const auto &[num, count] : counter)
            maxHeap.emplace(count, num);

        vector<int> result;
        for (int i = 0; i < k; ++i)
            result.push_back(maxHeap.top().second), maxHeap.pop();

        return result;
    }
};